crates/bevy_ecs/src/entity/map_entities.rs

```diff
@@ -3,46 +3,71 @@ use crate::{
     identifier::masks::{IdentifierMask, HIGH_MASK},
     world::World,
 };
+pub use bevy_ecs_macros::IterEntities;
 
 use super::EntityHashMap;
 
-/// Operation to map all contained [`Entity`] fields in a type to new values.
+/// Apply an operation to all entities in a container.
 ///
-/// As entity IDs are valid only for the [`World`] they're sourced from, using [`Entity`]
-/// as references in components copied from another world will be invalid. This trait
-/// allows defining custom mappings for these references via [`EntityMappers`](EntityMapper), which
-/// inject the entity mapping strategy between your `MapEntities` type and the current world
-/// (usually by using an [`EntityHashMap<Entity>`] between source entities and entities in the
-/// current world).
+/// This is implemented by default for types that implement [`IterEntities`] or
+/// where `&T` and `&mut T` both implement [`IntoIterator`].
 ///
-/// Implementing this trait correctly is required for properly loading components
-/// with entity references from scenes.
-///
-/// ## Example
-///
-/// ```
-/// use bevy_ecs::prelude::*;
-/// use bevy_ecs::entity::MapEntities;
+/// It may be useful to implement directly for types that can't produce an
+/// iterator for lifetime reasons, such as those involving internal mutexes.
+pub trait MapEntities {
+    /// Apply an operation to all contained entities.
+    fn map_entities<F: FnMut(Entity)>(&self, f: F);
```

Comment:
Do we have a use case for mapping entities without their mutation? Maybe remove this and just ask users to use `IterEntities` instead?

Reply:
I think there are situations where you can implement `map_entities` where you wouldn't be able to implement `iter_entities`, like when it involves a mutex lock or there's something else that needs to happen before/after iteration in the function body.

Reply:
I think in this case users will be able to create their own blanked trait impl since we now have `IterEntities`. I don't think non-mutable version for `MapEntities` is needed anymore.

Reply:
> I think in this case users will be able to create their own blanked trait impl since we now have IterEntities

The problem is that they _can't_ create their own trait/blanket impl for all `IterEntities` types, or rather that they can, but that not all types are able to have `IterEntities` implemented for them. In those cases, they'd have to specifically implement their trait for the desired types.

For example:

```rust
    struct SharedEntities {
        entities: Arc<Mutex<Vec<Entity>>>,
    }

    impl IterEntities for SharedEntities {
        fn iter_entities(&self) -> impl Iterator<Item = Entity> {
            self.entities.lock().unwrap().iter().copied()
        }

        fn iter_entities_mut(&mut self) -> impl Iterator<Item = &mut Entity> {
            self.entities.lock().unwrap().iter_mut()
        }
    }
```

gives you a 

```
    |
359 |         fn iter_entities(&self) -> impl Iterator<Item = Entity> {
    |                          - let's call the lifetime of this reference `'1`
360 |             self.entities.lock().unwrap().iter().copied()
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^----------------
    |             |
    |             creates a temporary value which is freed while still in use
    |             opaque type requires that borrow lasts for `'1`
361 |         }
    |         - temporary value is freed at the end of this statement
```

Wheras `MapEntities` doesn't have that problem. I think the workaround for returning an iterator over a type inside a mutex guard would involve an ugly self-referential struct to hold both the guard and the iterator referencing it.

Reply:
> The problem is that they can't create their own trait/blanket impl for all IterEntities types,

Why not?
I mean users can create something like `MapEntitiesVisitor` and impl it for all types that implement `IterEntities`.

Reply:
Alternatively we can rename it into `EntitiesVisitor`. Because `MapEntities` is no longer appropriate name.

Reply:
I guess I find `MapEntities` to still be appropriate in the functional sense :grin: 

But I'd be fine with `VisitEntities` instead.

>  I mean users can create something like MapEntitiesVisitor and impl it for all types that implement IterEntities.

But not all types that they may want `MapEntitiesVisitor` for _can_ implement `IterEntities`, so they'd have to do it manually. And then what if a third party wants to use a fourth-party type that the `MapEntitiesVisitor` author hasn't taken into account? Then they have to introduce wrapper structs to implement the trait on.

Reply:
> I guess I find MapEntities to still be appropriate in the functional sense 😁

:thinking: 
I would go with `VisitEntities`, yes :)